File: src/services/rag.ts
const API_BASE_URL = "http://localhost:8000";

export interface RAGSource {
  document: string;
  page: number;
  score: number;
}

export interface RAGAskResponse {
  answer: string;
  sources: RAGSource[];
}

export interface RAGIndexResponse {
  document_id: number;
  filename: string;
  chunks_indexed: number;
}

export async function indexDocument(documentId: number): Promise<RAGIndexResponse> {
  const res = await fetch(`${API_BASE_URL}/rag/index/${documentId}`, {
    method: "POST",
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: "Indexing failed" }));
    throw new Error(err.detail || "Indexing failed");
  }
  return res.json();
}

export async function askQuestion(question: string): Promise<RAGAskResponse> {
  const res = await fetch(`${API_BASE_URL}/rag/ask`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ question }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: "Query failed" }));
    throw new Error(err.detail || "Query failed");
  }
  return res.json();
}

File: backend/app/routes/rag.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException

from ..schemas.rag import RAGAskRequest, RAGAskResponse, RAGIndexResponse, RAGSource
from ..services.rag import generate_answer, index_document, register_document

router = APIRouter(prefix="/rag", tags=["rag"])


@router.post("/index/{document_id}", response_model=RAGIndexResponse)
async def index_uploaded_document(document_id: int):
    try:
        result = index_document(document_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Indexing failed: {str(e)}")
    return RAGIndexResponse(**result)


@router.post("/ask", response_model=RAGAskResponse)
async def ask_question(request: RAGAskRequest):
    if not request.question.strip():
        raise HTTPException(status_code=400, detail="Question cannot be empty")
    try:
        result = generate_answer(request.question)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Query failed: {str(e)}")
    return RAGAskResponse(
        answer=result["answer"],
        sources=[RAGSource(**s) for s in result["sources"]],
    )
